pkg/operator/encryption/state_controller_encryption_test.go

```diff
@@ -84,6 +84,304 @@ func TestEncryptionStateController(t *testing.T) {
 				}
 			},
 		},
+
+		// scenario 3
+		{
+			name:            "creates an EncryptionConfig with keys whose IDs are overflown",
+			targetNamespace: "kms",
+			destName:        "encryption-config-kube-apiserver-test",
+			targetGRs: map[schema.GroupResource]bool{
+				schema.GroupResource{Group: "", Resource: "secrets"}: true,
+			},
+			initialResources: []runtime.Object{
+				createDummyKubeAPIPod("kube-apiserver-1", "kms"),
+				createExpiredMigratedEncryptionKeySecretWithRawKey("kms", schema.GroupResource{"", "secrets"}, 100, []byte("61def964fb967f5d7c44a2af8dab6865")),
+				createWriteEncryptionKeySecretWithRawKey("kms", schema.GroupResource{"", "secrets"}, 101, []byte("171582a0fcd6c5fdb65cbf5a3e9249d7")),
+			},
+			expectedEncryptionCfg: func() *apiserverconfigv1.EncryptionConfiguration {
+				keys := []apiserverconfigv1.Key{
+					apiserverconfigv1.Key{
+						Name:   "1",
+						Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc=",
+					},
+					apiserverconfigv1.Key{
+						Name:   "0",
+						Secret: "NjFkZWY5NjRmYjk2N2Y1ZDdjNDRhMmFmOGRhYjY4NjU=",
+					},
+				}
+				ec := createEncryptionCfgWithWriteKey(keys, "secrets")
+				return ec
+			}(),
+			expectedActions: []string{"list:pods:kms", "list:secrets:openshift-config-managed", "get:secrets:openshift-config-managed", "create:secrets:openshift-config-managed", "create:events:kms"},
+			validateFunc: func(ts *testing.T, actions []clientgotesting.Action, destName string, expectedEncryptionCfg *apiserverconfigv1.EncryptionConfiguration) {
+				wasSecretValidated := false
+				for _, action := range actions {
+					if action.Matches("create", "secrets") {
+						createAction := action.(clientgotesting.CreateAction)
+						actualSecret := createAction.GetObject().(*corev1.Secret)
+						err := validateSecretWithEncryptionConfig(actualSecret, expectedEncryptionCfg, destName)
+						if err != nil {
+							ts.Fatalf("failed to verfy the encryption config, due to %v", err)
+						}
+						wasSecretValidated = true
+						break
+					}
+				}
+				if !wasSecretValidated {
+					ts.Errorf("the secret wasn't created and validated")
+				}
+			},
+		},
+
+		// scenario 4
+		{
+			name:            "secret with EncryptionConfig is created and it contains a single write key",
+			targetNamespace: "kms",
+			destName:        "encryption-config-kube-apiserver-test",
+			targetGRs: map[schema.GroupResource]bool{
+				schema.GroupResource{Group: "", Resource: "secrets"}: true,
+			},
+			initialResources: []runtime.Object{
+				createDummyKubeAPIPod("kube-apiserver-1", "kms"),
+				createReadEncryptionKeySecretWithRawKey("kms", schema.GroupResource{"", "secrets"}, 34, []byte("171582a0fcd6c5fdb65cbf5a3e9249d7")),
```

Comment:
a read-only key is promoted to a write key even if it doesn't have `write` annotation? what's the purpose of the `write` annotation then?

Reply:
> Once everyone observes that, it will be annotated as the write key, which will tell the migration code that storage migration is needed

Hmm, I think that the key doesn't have to be observed (annotated) as the write key to be considered by the migration code and I think that this test proves it https://github.com/enj/cluster-kube-apiserver-operator/pull/12/commits/3d0a2320afe70c1f0c5b3114be4d5592e7f5174b#diff-260dca05ba943635fbfb87b6cde4e826R31

Am I missing something?